internal/datastore/common/gc.go

```diff
@@ -0,0 +1,156 @@
+package common
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+var (
+	gcDurationHistogram = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "spicedb",
+		Subsystem: "datastore",
+		Name:      "gc_duration_seconds",
+		Help:      "The duration of datastore garbage collection.",
+		Buckets:   []float64{0.01, 0.1, 0.5, 1, 5, 10, 25, 60, 120},
+	})
+
+	gcRelationshipsClearedCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "spicedb",
+		Subsystem: "datastore",
+		Name:      "gc_relationships_total",
+		Help:      "The number of relationships cleared by the datastore garbage collection.",
+	})
+
+	gcTransactionsClearedCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "spicedb",
+		Subsystem: "datastore",
+		Name:      "gc_transactions_total",
+		Help:      "The number of transactions cleared by the datastore garbage collection.",
+	})
+
+	gcNamespacesClearedCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "spicedb",
+		Subsystem: "datastore",
+		Name:      "gc_namespaces_total",
+		Help:      "The number of namespaces cleared by the datastore garbage collection.",
+	})
+)
+
+// RegisterGCMetrics registers garbage collection metrics to the default
+// registry.
+func RegisterGCMetrics() error {
+	for _, metric := range []prometheus.Collector{
+		gcDurationHistogram,
+		gcRelationshipsClearedCounter,
+		gcTransactionsClearedCounter,
+		gcNamespacesClearedCounter,
+	} {
+		if err := prometheus.Register(metric); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GarbageCollector represents any datastore that supports external garbage
+// collection.
+type GarbageCollector interface {
+	IsReady(context.Context) (bool, error)
+	Now(context.Context) (time.Time, error)
+	TxIDBefore(context.Context, time.Time) (uint64, error)
+	DeleteBeforeTx(context.Context, uint64) (GarbageCollectedAmounts, error)
+}
+
+// GarbageCollectedAmounts is the collection of amounts of deletions from
+// an individual run of the garbage collector.
+type GarbageCollectedAmounts struct {
+	Relationships int64
```

Comment:
`RemovedRelationships`, etc?

Reply:
the callsites all now read `removed.Relationships`, which I think is a good middle-ground for verbosity